File: QID-3512-SFSLossDiscrete/SFSLossDiscrete.R
# Close all plots and clear variables
graphics.off()
rm(list = ls(all = TRUE))

# Plots of correlation dependence in discrete case

L     = c(0, 1, 2)
PD    = 0.2
RHO   = c(0, 0.2, 0.5, 1)
PL012 = c()

for (i in 1:length(RHO)) {
    rho   = RHO[i]
    PL2   = rho * (1 - PD) * PD + PD^2
    PL1   = PD * (1 - PD) * (1 - rho)
    PL0   = (1 - PD) - PD * (1 - PD) * (1 - rho)
    PL012 = cbind(PL012, c(PL0, PL1, PL2))
}

barplot(height = t(PL012), beside = TRUE, names.arg = c(0, 1, 2), col = c("dark blue", 
    "turquoise ", "yellow", "red"), xlab = "Loss", ylab = "Probability", legend.text = c("roh=0", 
    "roh=0.2", "roh=0.5", "roh=1")) 

File: QID-3512-SFSLossDiscrete/SFSLossDiscrete.m
% ----------------------------------------------------------------------
% Book:         SFS
% ----------------------------------------------------------------------
% Quantlet:     SFSLossDiscrete
% ----------------------------------------------------------------------
% Description:  Loss distribution of portfolio of two zero coupon bonds
%               with zero recovery.
%               Corresponds to excercise 18.3 in SFS.
% ----------------------------------------------------------------------
% Usage:        -
% ----------------------------------------------------------------------
% Inputs:       PD  - probability of default
%               RHO - correlation
% ----------------------------------------------------------------------
% Output:       plot of the portfolio loss distribution for four different
%               correlation paramenters.
% ----------------------------------------------------------------------
% Example:      the example is produced for the values:
%               PD = 0.2
%               RHO  = [0,0.2,0.5,1]
% -----------------------------------------------------------------------
% Author:       Szymon Borak 20090731
% -----------------------------------------------------------------------

close all
clear
clc

%Plots of correlation dependence in discrete case

L     = [0,1,2];
PD    = 0.2;
RHO   = [0,0.2,0.5,1];
PL012 = [];

for i =1:length(RHO)
    rho   = RHO(i);
    PL2   = rho*(1-PD)*PD + PD^2;
    PL1   = PD*(1-PD)*(1-rho);
    PL0   = (1-PD)-PD*(1-PD)*(1-rho);
    PL012 = [PL012,[PL0;PL1;PL2]];
end

bar(L,PL012,'group')
legend( ['\rho =' num2str(RHO(1))],['\rho=' num2str(RHO(2))],['\rho=' num2str(RHO(3))],['\rho = ' num2str(RHO(4))])

xlabel('Loss')
ylabel('Probability')
